### PR TITLE
provider/aws: Add JSON validation to the aws_cloudwatch_event_rule resource.

### PR DESCRIPTION
**Pull request by kwilczynski**
This commit adds support for new helper function which is used to
normalise and validate JSON string. Also adds new validation function
to wrap around the aforementioned helper.

Signed-off-by: Krzysztof Wilczynski <krzysztof.wilczynski@linux.com>

### CODE REVIEW

**Comment by kwilczynski** [comment]
Part of work which began in #8028.

**Comment by kwilczynski** [comment]
Tests are passing:

(unit)

```
$ make test TEST=./builtin/providers/aws TESTARGS='-v -run=TestResourceAWSCloudWatchEventRule_validateEventPatternValue'
==> Checking that code complies with gofmt requirements...
==> Checking AWS provider for unchecked errors...
==> NOTE: at this time we only look for uncheck errors in the AWS package
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/09/14 22:40:23 Generated command/internal_plugin_list.go
TF_ACC= go test ./builtin/providers/aws -v -run=TestResourceAWSCloudWatchEventRule_validateEventPatternValue -timeout=30s -parallel=4
=== RUN   TestResourceAWSCloudWatchEventRule_validateEventPatternValue
--- PASS: TestResourceAWSCloudWatchEventRule_validateEventPatternValue (0.00s)
PASS
ok     	github.com/hashicorp/terraform/builtin/providers/aws   	0.021s
```

(acceptance)

```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSCloudWatchEventRule'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/09/14 19:36:21 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSCloudWatchEventRule -timeout 120m
=== RUN   TestAccAWSCloudWatchEventRule_importBasic
--- PASS: TestAccAWSCloudWatchEventRule_importBasic (15.01s)
=== RUN   TestAccAWSCloudWatchEventRule_basic
--- PASS: TestAccAWSCloudWatchEventRule_basic (25.89s)
=== RUN   TestAccAWSCloudWatchEventRule_full
--- PASS: TestAccAWSCloudWatchEventRule_full (15.08s)
=== RUN   TestAccAWSCloudWatchEventRule_enable
--- PASS: TestAccAWSCloudWatchEventRule_enable (37.42s)
PASS
ok     	github.com/hashicorp/terraform/builtin/providers/aws   	93.441s
Raw
```

File: builtin/providers/aws/resource_aws_cloudwatch_event_rule.go [hunk 1]
```diff
@@ -247,3 +252,17 @@ func getStringStateFromBoolean(isEnabled bool) string {
...
+func validateEventPatternValue(length int) schema.SchemaValidateFunc {
+	return func(v interface{}, k string) (ws []string, errors []error) {
+		value := v.(string)
+		if len(value) > length {
+			errors = append(errors, fmt.Errorf(
+				"%q cannot be longer than %d characters: %q", k, length, value))
+		}
+		if _, err := normalizeJsonString(value); err != nil {
+			errors = append(errors, fmt.Errorf("%q contains an invalid JSON: %s", k, err))
+		}
```
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_cloudwatch_event_rule.go]: We send the **minified** (normalized) JSON to the API which is the one that is limited to 2048 characters, not the original JSON string with spaces, newlines etc.

At this point we would prevent the user from sending minified JSON to the API which can be below 2048 chars after stripping whitespaces etc. Also I'm not sure if there is any point in checking length of invalid JSON.

It may be worth covering this particular edge case by a test.

**Comment by radeksimko** [comment]
I'm happy to merge this once the mentioned edge case w/ normalization & length check is addressed - this otherwise looks good.

**Comment by kwilczynski** [comment]
@radeksimko ah, that is a very good point! I will fix this right away.

**Comment by kwilczynski** [comment]
Unit test is passing:

```
 make test TEST=./builtin/providers/aws TESTARGS='-v -run=TestResourceAWSCloudWatchEventRule_validateEventPatternValue'
==> Checking that code complies with gofmt requirements...
==> Checking AWS provider for unchecked errors...
==> NOTE: at this time we only look for uncheck errors in the AWS package
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/09/18 15:05:26 Generated command/internal_plugin_list.go
TF_ACC= go test ./builtin/providers/aws -v -run=TestResourceAWSCloudWatchEventRule_validateEventPatternValue -timeout=30s -parallel=4
=== RUN   TestResourceAWSCloudWatchEventRule_validateEventPatternValue
--- PASS: TestResourceAWSCloudWatchEventRule_validateEventPatternValue (0.00s)
PASS
ok     	github.com/hashicorp/terraform/builtin/providers/aws   	0.015s
```

**Comment by kwilczynski** [comment]
@radeksimko I have updated the `ValidateFunc`, it should do the right thing now. Thank you!
